query-engine/connectors/sql-query-connector/src/query_builder/select/lateral.rs

```diff
@@ -6,13 +6,22 @@ use crate::{
     model_extensions::AsColumn,
 };
 
+use std::collections::HashMap;
+
 use quaint::ast::*;
 use query_structure::*;
 
 /// Select builder for joined queries. Relations are resolved using LATERAL JOINs.
 #[derive(Debug, Default)]
 pub(crate) struct LateralJoinSelectBuilder {
     alias: Alias,
+
+    // TODO: consider using references to `VirtualSelection` as keys to avoid cloning. This would
+    // have an effect on a big chunk of API of `JoinSelectBuilder` as we will need to prove that
+    // the builder itself does not outlive the `FieldSelection` passed as the parameter of `build`
+    // method, and introduce relationships between the lifetime of `self` and parameters of various
+    // methods.
+    visited_virtuals: HashMap<VirtualSelection, String>,
```

Comment:
There's also an edge case I now understand I hadn't considered: what if a query shuttles back and forth between one-to-many and many-to-one fields a few times and selects a relation count for the same relation on different levels.

I'm not sure without checking or thinking more if this will cause problems, but I'll check and add tests. If it is indeed problematic, then we'll need to either:

* change the data structure, either
  - by making the nesting level part of the key, or
  - by making the value a stack, or
* change the algorithm to not rely on this state at all: bring back `relation_count_alias_name` for lateral join builder, and make `with_virtual_selections` and `build_json_obj_virtual_selection` independently decide whether any relation query _would have_ processed a given virtual field rather than rely on a record that it _did_.

Reply:
It seems to work fine as it is because the implicit call stack together with the fact that we already remove the alias from the hash map avoid conflicts in a similar way as an explicit stack stored in the data structure would, but I ran into a different bug with this pattern that is also reproducible on main: `Field '_count' is not a model field and doesn't have a supported type for a virtual field"`. The generated query looks correct so I'll have to debug what the problem is.

Reply:
>  ran into a different bug with this pattern
The problem was not the pattern but having `_count` selected before any other fields. I'll open an issue about it.

Reply:
https://github.com/prisma/team-orm/issues/927

Reply:
and tests added